# 01. Import libraries

In [3]:
import pandas as pd
import numpy as np
import os

# 02. Import data

In [6]:
path = '/Users/bobbi/Careerfoundry/InstacartBasketAnalysis_01_25'
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original data', 'products.csv'), index_col = False)
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared data', 'orders_wrangled.csv'), index_col = False)

# 03. Mixed-Type Data

In [8]:
df_ords.describe()

,Unnamed: 0,order_id,user_id,order_number,order_dow,order_hour_of_day,days_since_last_order
count,3.421083e+06,3.421083e+06,3.421083e+06,3.421083e+06,3.421083e+06,3.421083e+06,3.214874e+06
mean,1.710541e+06,1.710542e+06,1.029782e+05,1.715486e+01,2.776219e+00,1.345202e+01,1.111484e+01
std,9.875817e+05,9.875817e+05,5.953372e+04,1.773316e+01,2.046829e+00,4.226088e+00,9.206737e+00
min,0.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
25%,8.552705e+05,8.552715e+05,5.139400e+04,5.000000e+00,1.000000e+00,1.000000e+01,4.000000e+00
50%,1.710541e+06,1.710542e+06,1.026890e+05,1.100000e+01,3.000000e+00,1.300000e+01,7.000000e+00
75%,2.565812e+06,2.565812e+06,1.543850e+05,2.300000e+01,5.000000e+00,1.600000e+01,1.500000e+01
max,3.421082e+06,3.421083e+06,2.062090e+05,1.000000e+02,6.000000e+00,2.300000e+01,3.000000e+01


In [9]:
# Create a dataframe
df_test = pd.DataFrame()

In [10]:
# Create a mixd type column 
df_test['mix'] = ['a', 'b', 1, True]

In [11]:
df_test.head()

,mix
0,a
1,b
2,1
3,True


In [12]:
# Check for mixed types

for col in df_test.columns:
    # Get the reference type from the first element of the column
    ref_type = type(df_test[col].iloc[0])
    
    # Check if any element in the column has a type different from the reference type
    weird = df_test[col].apply(type) != ref_type
    
    # If there are any mismatched types, print the column name
    if weird.any():
        print(col)

mix


In [13]:
df_test['mix'] = df_test['mix'].astype('str')

In [14]:
df_test

,mix
0,a
1,b
2,1
3,True


In [15]:
df_test.dtypes

mix    object
dtype: object

# 04. Missing values

In [17]:
# Find missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [18]:
# Print rows with null values in 'product_name' column
df_nan = df_prods[df_prods['product_name'].isnull() == True]

df_nan

,product_id,product_name,aisle_id,department_id,prices
33,34,NaN,121,14,12.2
68,69,NaN,26,7,11.8
115,116,NaN,93,3,10.8
261,262,NaN,110,13,12.1
525,525,NaN,109,11,1.2
1511,1511,NaN,84,16,14.3
1780,1780,NaN,126,11,12.3
2240,2240,NaN,52,1,14.2
2586,2586,NaN,104,13,12.4
3159,3159,NaN,126,11,13.1


In [19]:
# Eliminate null values and create new dataframe
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

df_prods_clean.shape

(49677, 5)

# 05. Duplicates

In [21]:
# Finding duplicates

df_dups = df_prods_clean[df_prods_clean.duplicated()]
df_dups

,product_id,product_name,aisle_id,department_id,prices
462,462,Fiber 4g Gummy Dietary Supplement,70,11,4.8
18459,18458,Ranger IPA,27,5,9.2
26810,26808,Black House Coffee Roasty Stout Beer,27,5,13.4
35309,35306,Gluten Free Organic Peanut Butter & Chocolate ...,121,14,6.8
35495,35491,Adore Forever Body Wash,127,11,9.9


In [22]:
# Drop duplicates 
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

df_prods_clean_no_dups.shape

(49672, 5)

In [37]:
# Exporting data
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# 06. Exercises

### Consistency check for df_ords

In [26]:
df_ords.describe()

,Unnamed: 0,order_id,user_id,order_number,order_dow,order_hour_of_day,days_since_last_order
count,3.421083e+06,3.421083e+06,3.421083e+06,3.421083e+06,3.421083e+06,3.421083e+06,3.214874e+06
mean,1.710541e+06,1.710542e+06,1.029782e+05,1.715486e+01,2.776219e+00,1.345202e+01,1.111484e+01
std,9.875817e+05,9.875817e+05,5.953372e+04,1.773316e+01,2.046829e+00,4.226088e+00,9.206737e+00
min,0.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
25%,8.552705e+05,8.552715e+05,5.139400e+04,5.000000e+00,1.000000e+00,1.000000e+01,4.000000e+00
50%,1.710541e+06,1.710542e+06,1.026890e+05,1.100000e+01,3.000000e+00,1.300000e+01,7.000000e+00
75%,2.565812e+06,2.565812e+06,1.543850e+05,2.300000e+01,5.000000e+00,1.600000e+01,1.500000e+01
max,3.421082e+06,3.421083e+06,2.062090e+05,1.000000e+02,6.000000e+00,2.300000e+01,3.000000e+01


In [27]:
# Change column name to keep consistency and clarity 
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

df_ords.head()

,Unnamed: 0,order_id,user_id,eval_set,order_number,orders_day_of_week,order_hour_of_day,days_since_last_order
0,0,2539329,1,prior,1,2,8,NaN
1,1,2398795,1,prior,2,3,7,15.0
2,2,473747,1,prior,3,3,12,21.0
3,3,2254736,1,prior,4,4,7,29.0
4,4,431534,1,prior,5,4,15,28.0


In [28]:
# Drop unnamed column as it seems to be a duplicate index
df_ords.drop(columns='Unnamed: 0', inplace=True)

df_ords

,order_id,user_id,eval_set,order_number,orders_day_of_week,order_hour_of_day,days_since_last_order
0,2539329,1,prior,1,2,8,NaN
1,2398795,1,prior,2,3,7,15.0
2,473747,1,prior,3,3,12,21.0
3,2254736,1,prior,4,4,7,29.0
4,431534,1,prior,5,4,15,28.0
...,...,...,...,...,...,...,...
3421078,2266710,206209,prior,10,5,18,29.0
3421079,1854736,206209,prior,11,4,10,30.0
3421080,626363,206209,prior,12,1,12,18.0
3421081,2977660,206209,prior,13,1,12,7.0


In [32]:
df_ords.describe()

,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_last_order
count,3.421083e+06,3.421083e+06,3.421083e+06,3.421083e+06,3.421083e+06,3.214874e+06
mean,1.710542e+06,1.029782e+05,1.715486e+01,2.776219e+00,1.345202e+01,1.111484e+01
std,9.875817e+05,5.953372e+04,1.773316e+01,2.046829e+00,4.226088e+00,9.206737e+00
min,1.000000e+00,1.000000e+00,1.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
25%,8.552715e+05,5.139400e+04,5.000000e+00,1.000000e+00,1.000000e+01,4.000000e+00
50%,1.710542e+06,1.026890e+05,1.100000e+01,3.000000e+00,1.300000e+01,7.000000e+00
75%,2.565812e+06,1.543850e+05,2.300000e+01,5.000000e+00,1.600000e+01,1.500000e+01
max,3.421083e+06,2.062090e+05,1.000000e+02,6.000000e+00,2.300000e+01,3.000000e+01


If we pay attention to the orders_day_of_weel we can see that days are counted from 0-6 instead of 1-7 but makes sense if goes grom Sunday = 0 to Saturday = 6 for example.
In order_hour_of_day the range is 0-23 which makes sense for a 24hours day.
In days_since_last_order the range goes from minimum 0 to maximum 30 fitting for a month cycle count.

The other columns seem normal, order_id has large range but it is normas as it represents the number of orders.

### Check for mixed-type data in your df_ords dataframe.

In [49]:
for col in df_ords.columns:
    ref_type = type(df_ords[col].iloc[0])            #get reference type from first column
    weird = df_ords[col].apply(type) != ref_type     #check if there is any type different from ref_type
    if weird.any():                                 #if any different type, print that column
        print(col)

order_id
user_id
order_number
orders_day_of_week
order_hour_of_day
days_since_last_order


In [52]:
df_ords['order_id'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 3421083 entries, 0 to 3421082
Series name: order_id
Non-Null Count    Dtype
--------------    -----
3421083 non-null  int64
dtypes: int64(1)
memory usage: 26.1 MB


In [58]:
df_ords[['order_id', 'user_id', 'order_number', 'orders_day_of_week', 'order_hour_of_day', 'days_since_last_order']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 6 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   order_id               int64  
 1   user_id                int64  
 2   order_number           int64  
 3   orders_day_of_week     int64  
 4   order_hour_of_day      int64  
 5   days_since_last_order  float64
dtypes: float64(1), int64(5)
memory usage: 156.6 MB


I couldn't use the following loop cause .applymap() is depricated
```
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)
```
I tried with other loop comparing the datatypes at the first row but it didn't work because the first row might be NaN (whose type is float) and therefore it prints all the columns.

In [67]:
# Using a variable from the original loop 
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

I asked chatgpt to provide an alternative, as my orginal idea didn't work, and provided the code above.
It seems there are no columns with mixed types.

### Run a check for missing values in your df_ords dataframe.

In [72]:
df_ords.isnull().sum()

order_id                      0
user_id                       0
eval_set                      0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

There are 206209 missing values in days_since_last_order. It represent more than 10% of the data, it represents valueable information that cannot be deleted.

One explanation could be that first orders don't have any value in that column as it represent the time/days between orders.

### Run a check for missing values in your df_ords dataframe.

To address properly the missing values:
* First I am going to create a dataframe for the NAN orders.
* Once that we have all in the same dataframe I will check if the order_number is 1 for all of them and check if all have missing values for days_since_last order.
* At last create an extra boolean column to specify if it's the first purchase or not.

In [77]:
df_ords_nan = df_ords[df_ords['days_since_last_order'].isnull() == True]

df_ords_nan

,order_id,user_id,eval_set,order_number,orders_day_of_week,order_hour_of_day,days_since_last_order
0,2539329,1,prior,1,2,8,NaN
11,2168274,2,prior,1,2,11,NaN
26,1374495,3,prior,1,1,14,NaN
39,3343014,4,prior,1,6,11,NaN
45,2717275,5,prior,1,3,12,NaN
...,...,...,...,...,...,...,...
3420930,969311,206205,prior,1,4,12,NaN
3420934,3189322,206206,prior,1,3,18,NaN
3421002,2166133,206207,prior,1,6,19,NaN
3421019,2227043,206208,prior,1,1,15,NaN


In [92]:
# Check if all NAN have order_number = 1
if (df_ords_nan['order_number'] == 1).all():
    print("All missing values first orders.")
else:
    print("Missing values are not first orders.")

All missing values first orders.


In [95]:
# Create an extra boolean column for first orders.

df_ords['is_first_order'] = df_ords['days_since_last_order'].isnull()

df_ords.head()

,order_id,user_id,eval_set,order_number,orders_day_of_week,order_hour_of_day,days_since_last_order,is_first_order
0,2539329,1,prior,1,2,8,NaN,True
1,2398795,1,prior,2,3,7,15.0,False
2,473747,1,prior,3,3,12,21.0,False
3,2254736,1,prior,4,4,7,29.0,False
4,431534,1,prior,5,4,15,28.0,False


Once we have the extra column another extra step could be to substitute the NAN values for 0 values and avoid having missing values in our column
```
    df_ords['days_since_last_order'].fillna(0, inplace=True)
```

But for now I rather leave them as NAN.

### Run a check for duplicate values in your df_ords data.

In [101]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [103]:
df_ords_dups

,order_id,user_id,eval_set,order_number,orders_day_of_week,order_hour_of_day,days_since_last_order,is_first_order


The new dataframe had no full duplicates.

### Export data

In [107]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))